File: python/CarFleet.py
from typing import List
class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        fleet = []
        stack = []

        for i in range(len(position)):
            fleet.append([position[i], speed[i]])
        
        fleet = (sorted(fleet, key = lambda x: x[0]))[::-1]

        # if the car time is less than the first car, then it means that they become a single fleet and we can pop it from a stack
        for car in fleet:
            carTime = (target - car[0]) / car[1]
            stack.append(carTime)
            if len(stack) >= 2 and stack[-1] <= stack[-2]:
                stack.pop()

        return len(stack)

File: cpp/CarFleet.cpp
#include <algorithm>
#include <iostream>
#include <vector>
#include <stack>

using namespace std;
class Solution {
public:
    int carFleet(int target, vector<int>& position, vector<int>& speed) {
        
        stack<float> stack;
        vector<pair<int, float>> fleet;
         
        for(int i = 0; i < position.size(); i++)
            fleet.push_back({ position[i], float(target - position[i]) / speed[i] });
        
        sort(fleet.begin(), fleet.end());
                   
        
        for(int i = 0; i < position.size(); i++){
            float carTime = fleet[i].second;
         
            while(stack.size() && (carTime >= stack.top()))
                stack.pop();
            
            stack.push(carTime);
        }
        
        return stack.size();
    }
};